File: KotlinMobileApp/Java/Start/app/src/main/java/com/rachelleboyette/catquiz/viewmodels/CatFactViewModel.java
package com.rachelleboyette.catquiz.viewmodels;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.rachelleboyette.catquiz.models.CatFact;
import com.rachelleboyette.catquiz.services.CatApiService;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableSingleObserver;

public class CatFactViewModel extends ViewModel {
    private CatApiService catApiService = new CatApiService();
    private MutableLiveData<String> catFactText = new MutableLiveData<>();
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public CatFactViewModel() {
        callCatFact();
    }

    private void callCatFact() {
        Disposable apiSubscription = catApiService.loadCatFact().subscribeWith(new DisposableSingleObserver<CatFact>() {
            @Override
            public void onSuccess(CatFact catFact) {
                publishCatFact(catFact);
            }

            @Override
            public void onError(Throwable e) {
                onApiError(e);
            }
        });

        compositeDisposable.add(apiSubscription);
    }

    private void onApiError(Throwable e) {

    }

    private void publishCatFact(CatFact catFact) {

    }
}

File: KotlinMobileApp/KotlinMobileApp/app/src/main/java/com/neudesic/catquiz/viewmodels/CatFactViewModel.kt
package com.neudesic.catquiz.viewmodels

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.neudesic.catquiz.services.CatApiService
import com.neudesic.catquiz.models.CatFact
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.subscribeBy

class CatFactViewModel: ViewModel() {
    private var catApiService: CatApiService = CatApiService()
    var catFactText: MutableLiveData<String> = MutableLiveData()
    private var apiSubscription : CompositeDisposable = CompositeDisposable()

     init {
         callCatFact()
    }

    private fun publishCatFact(catFact: CatFact?) {
        catFactText.postValue(catFact?.fact)
    }

    private fun callCatFact() {
        val apiObserver = catApiService.loadCatFact().subscribeBy (
            onError = {
                onApiError(it)
            },
            onSuccess = {
                publishCatFact(it)
            }
        )

        apiSubscription.add(apiObserver)
    }

    private fun onApiError(error: Throwable) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}